### PR TITLE
[SYCL] Add graph record and replay entry points

### PR DESCRIPTION
**Pull request by Bensuo**
- Add entry points to queue for graph record and replay (empty implementation)
- Add queue_state enum (currently unused)
- Move graph_state enum into separate header.

Note: Targeting the sycl-graph-poc-v2 branch for now until #5 is merged.

### CODE REVIEW

File: sycl/include/sycl/queue.hpp [hunk 1]
```diff
@@ -1068,6 +1083,22 @@ class __SYCL_EXPORT queue {
   /// \return the backend associated with this queue.
   backend get_backend() const noexcept;
 
+public:
+  /// Places the queue into command_graph recording mode.
+  /// Returns true if the queue was not already in recording mode.
+  bool begin_recording(
+      sycl::ext::oneapi::experimental::command_graph<
```
Comment (EwanC) [inline sycl/include/sycl/queue.hpp]: we might be able to drop the `sycl::` part of the namespace from the start of this type (and others you've added)

**Review by EwanC — APPROVED** [review]
LGTM, minor comment

File: sycl/include/sycl/queue.hpp (hunk 1 above)
Comment (Bensuo) [reply sycl/include/sycl/queue.hpp]: Done!

File: sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+namespace oneapi {
+namespace experimental {
+enum class graph_state {
+  modifiable,
+  executable,
+};
+}
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
```
Comment (reble) [inline sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp]:
```suggestion
} // namespace sycl

```

File: sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#pragma once
+
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+enum class graph_state {
+  modifiable,
+  executable,
+};
```
Comment (reble) [inline sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp:20]:
```suggestion
};
/// Forward declare graphs
template <graph_state state> class command_graph;
```

File: sycl/include/sycl/queue.hpp [hunk 4]
```diff
@@ -51,6 +53,16 @@
...
+/// Forward declare graphs
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+template <graph_state state> class command_graph;
+}
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
```
Comment (reble) [inline sycl/include/sycl/queue.hpp]:
```suggestion
```

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 5]
```diff
@@ -9,12 +9,18 @@
...
+#include "graph_defines.hpp"
 
 #include <list>
 #include <set>
 
 //__SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+class queue;
+}
```
Comment (reble) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp]:
```suggestion
} // namespace sycl
```

**Review by reble — APPROVED** [review]
Changes look good to me. Only a few suggestions.

File: sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp (hunk 2 above)
Comment (Bensuo) [reply sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp]: Fixed thanks!

File: sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp (hunk 3 above)
Comment (Bensuo) [reply sycl/include/sycl/ext/oneapi/experimental/graph_defines.hpp:20]: Good call, done!

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 5 above)
Comment (Bensuo) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: Good catch! I think this should be the inline namespace so I've added that instead.

**Comment by Bensuo** [comment]
Small update: I've add a simple test for record and replay alongside the existing explicit tests and I've also put the graphs code into the inline V1 namespace. I was having issues with linking the new queue functions when actually building against code that used them and this solved it, but I'm not sure I fully understand why it was causing issues in the first place. 

It does make sense to me to have it in the V1 namespace like the rest of the codebase for ABI reasons unless there's any objections @EwanC @reble

File: sycl/test/graph/graph-record-simple.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+
+#include <iostream>
+#include <sycl/sycl.hpp>
+
+#include <sycl/ext/oneapi/experimental/graph.hpp>
+
+const size_t n = 10;
```
Comment (EwanC) [inline sycl/test/graph/graph-record-simple.cpp]: this doesn't need to be a global variable

File: sycl/test/graph/graph-record-simple.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  sycl::property_list properties{
+      sycl::property::queue::in_order(),
+      sycl::ext::oneapi::property::queue::lazy_execution{}};
+
+  sycl::queue q{sycl::gpu_selector_v, properties};
+
+  sycl::ext::oneapi::experimental::command_graph g;
+
+  float *arr = sycl::malloc_shared<float>(n, q);
```
Comment (EwanC) [inline sycl/test/graph/graph-record-simple.cpp:20]: Should probably `sycl::free` this at the end

**Review by EwanC — APPROVED** [review]

File: sycl/test/graph/graph-record-simple.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  auto exec_graph = g.finalize(q.get_context());
+
+  auto result_before_exec2 = arr[0];
+
+  q.submit(exec_graph);
+
+  auto result = arr[0];
+
+  std::cout << "done.\n";
```
Comment (julianmi) [inline sycl/test/graph/graph-record-simple.cpp:48]: Should we check for the correct results?

File: sycl/test/graph/graph-record-simple.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+const size_t n = 10;
+
+int main() {
+
+  sycl::property_list properties{
+      sycl::property::queue::in_order(),
+      sycl::ext::oneapi::property::queue::lazy_execution{}};
+
+  sycl::queue q{sycl::gpu_selector_v, properties};
```
Comment (julianmi) [inline sycl/test/graph/graph-record-simple.cpp]: Should we use the default device selector to enable this test on other devices as well?

**Review by julianmi — APPROVED** [review]
LGTM. I have added two minor comments.

File: sycl/test/graph/graph-record-simple.cpp (hunk 6 above)
Comment (Bensuo) [reply sycl/test/graph/graph-record-simple.cpp]: Fixed!

File: sycl/test/graph/graph-record-simple.cpp (hunk 7 above)
Comment (Bensuo) [reply sycl/test/graph/graph-record-simple.cpp:20]: Done thanks!

File: sycl/test/graph/graph-record-simple.cpp (hunk 8 above)
Comment (Bensuo) [reply sycl/test/graph/graph-record-simple.cpp:48]: Good point, I've added a check for correct results now!

File: sycl/test/graph/graph-record-simple.cpp (hunk 9 above)
Comment (Bensuo) [reply sycl/test/graph/graph-record-simple.cpp]: Done!

**Pull request merged**